rust/src/lockfile.rs

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright (C) 2018 Red Hat, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+use c_utf8::CUtf8Buf;
+use serde_derive::{Deserialize, Serialize};
+use serde_json;
+use serde_yaml;
+use std::collections::HashMap;
+use std::path::Path;
+use std::{fs, io};
+use failure::Fallible;
+use failure::ResultExt;
+
+pub struct Lockfile {
+    #[allow(dead_code)] // Not used in tests
+    parsed: LockfileConfig,
+    serialized: CUtf8Buf
+}
+
+#[derive(PartialEq)]
+enum InputFormat {
+    YAML,
+    JSON,
```

Comment:
Since lockfiles are not written (or normally read) by humans, let's just make them JSON unconditionally?

Reply:
And once we do that...it seems to me the Rust code here isn't doing a whole lot for us, we end up re-parsing via `g_autoptr(GVariant) jsonmetav = json_gvariant_deserialize (metarootval, "{sa{sa(ss)}}", error);`.

The error messages from that are probably worse than serde, but eh...not sure it's worth carrying the Rust just for that.

On the other hand we could try to do more of the parsing Rust side.

Reply:
Ok. I'll see if I can figure out how to pass a hash map from rust to C.

Reply:
Why do we need to do that?  I'm saying we wouldn't use any Rust at all, just open the JSON file on the C side via json-glib directly.

Reply:
It would be cool to use more Rust of course but I am not seeing a way to avoid having to at least create minimal stubs for the libdnf API.  That would be a valuable thing to do but also not trivial.

Reply:
Ok

Reply:
Not using rust anymore.